crates/sui-graphql-client/src/faucet.rs

```diff
@@ -0,0 +1,235 @@
+
+use crate::{DEVNET_HOST, LOCAL_HOST};
+
+const FAUCET_DEVNET_HOST: &str = "https://faucet.devnet.sui.io/v1/gas";
+const FAUCET_TESTNET_HOST: &str = "https://faucet.testnet.sui.io/v1/gas";
+const FAUCET_LOCAL_HOST: &str = "http://localhost:9123/v1/gas";
+
+const FAUCET_REQUEST_TIMEOUT: Duration = Duration::from_secs(120);
+const FAUCET_POLL_INTERVAL: Duration = Duration::from_secs(2);
+
+pub struct FaucetClient {
+    faucet_url: Url,
+    inner: reqwest::Client,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct BatchFaucetResponse {
+    // This string is the Uuid for the req
+    pub task: Option<String>,
+    pub error: Option<String>,
+}
+
+#[derive(serde::Deserialize)]
+pub struct FaucetResponse {
+    task: Option<String>,
+    error: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct BatchStatusFaucetResponse {
+    pub status: Option<BatchSendStatus>,
+    pub error: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[serde(rename_all = "UPPERCASE")]
+pub enum BatchSendStatusType {
+    Inprogress,
+    Succeeded,
+    Discarded,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BatchSendStatus {
+    pub status: BatchSendStatusType,
+    pub transferred_gas_objects: Option<FaucetReceipt>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct FaucetReceipt {
+    pub sent: Vec<CoinInfo>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BatchFaucetReceipt {
+    pub task: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct CoinInfo {
+    pub amount: u64,
+    pub id: ObjectId,
+    pub transfer_tx_digest: TransactionDigest,
+}
+
+impl FaucetClient {
+    /// Construct a new `FaucetClient` with the given `reqwest::Client`. Defaults to Testnet
+    /// faucet.
+    pub(crate) fn new(inner: reqwest::Client, rpc: &Url) -> Self {
+        let faucet_url = if rpc.as_str() == LOCAL_HOST {
+            Url::parse(FAUCET_LOCAL_HOST).expect("Invalid faucet URL")
+        } else if rpc.as_str() == DEVNET_HOST {
+            Url::parse(FAUCET_DEVNET_HOST).expect("Invalid faucet URL")
+        } else {
+            Url::parse(FAUCET_TESTNET_HOST).expect("Invalid faucet URL")
+        };
+        FaucetClient { faucet_url, inner }
+    }
+
+    /// Set to local faucet.
+    pub fn local(mut self) -> Self {
+        self.faucet_url = Url::parse(FAUCET_LOCAL_HOST).expect("Invalid faucet URL");
+        self
+    }
+
+    /// Set to devnet faucet.
+    pub fn devnet(mut self) -> Self {
+        self.faucet_url = Url::parse(FAUCET_DEVNET_HOST).expect("Invalid faucet URL");
+        self
+    }
+
+    /// Set to testnet faucet.
+    pub fn testnet(mut self) -> Self {
+        self.faucet_url = Url::parse(FAUCET_TESTNET_HOST).expect("Invalid faucet URL");
+        self
+    }
```

Comment:
Lets remove these methods that mutate the client and instead have them be constructors for network specific faucets

Reply:
Fixed.